clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -278,38 +306,102 @@
+                        // log a warning and skip this record
+                        log.warn("Partition '{}' is assigned to multiple consumers following sticky assignment generation {}.",
+                                partition, consumerUserData.generation);
+                    } else
+                        consumers.put(consumerUserData.generation, consumer);
+                } else {
+                    Map<Integer, String> sortedConsumers = new TreeMap<>();
+                    sortedConsumers.put(consumerUserData.generation, consumer);
+                    sortedPartitionConsumersByGeneration.put(partition, sortedConsumers);
+                }
+            }
+        }
+
+        // prevAssignment holds the prior ConsumerGenerationPair (before current) of each partition
+        // current and previous consumers are the last two consumers of each partition in the above sorted map
+        for (Map.Entry<TopicPartition, Map<Integer, String>> partitionConsumersEntry: sortedPartitionConsumersByGeneration.entrySet()) {
+            TopicPartition partition = partitionConsumersEntry.getKey();
+            TreeMap<Integer, String> consumers = (TreeMap<Integer, String>) partitionConsumersEntry.getValue();
+            Iterator<Integer> it = consumers.descendingKeySet().iterator();
+
+            // let's process the current (most recent) consumer first
+            String consumer = consumers.get(it.next());
+            if (currentAssignment.containsKey(consumer))
+                currentAssignment.get(consumer).add(partition);
+            else {
+                List<TopicPartition> partitions = new ArrayList<>(Collections.singletonList(partition));
+                currentAssignment.put(consumer, partitions);
+            }
+
+            // now update previous assignment if any
+            if (it.hasNext()) {
+                int generation = it.next();
+                prevAssignment.put(partition, new ConsumerGenerationPair(consumers.get(generation), generation));
+            }
+            it.remove();
         }
     }
 
     @Override
     public void onAssignment(Assignment assignment) {
         memberAssignment = assignment.partitions();
+        generation = deserializeTopicPartitionAssignment(assignment.userData()).generation;
     }
 
     @Override
     public Subscription subscription(Set<String> topics) {
         if (memberAssignment == null)
             return new Subscription(new ArrayList<>(topics));
 
-        return new Subscription(new ArrayList<>(topics), serializeTopicPartitionAssignment(memberAssignment));
+        return new Subscription(new ArrayList<>(topics),
+                serializeTopicPartitionAssignment(new ConsumerUserData(memberAssignment, generation)));
     }
 
     @Override
     public String name() {
         return "sticky";
     }
 
+    @Override
+    protected Map<String, Assignment> getAssignments(Map<String, List<TopicPartition>> rawAssignments) {
```

Comment:
I am a bit confused about this. Previously we weren't defining user data for the assignment. We only needed to provide the old assignment in the subscription user data. Was that not sufficient?